Merge main branch into Ogre 2.2 and fixes

darksylinc opened the pull request:
# 🦟 Bug fix

 - Merged main branch into Ogre 2.2 branch
     - Rather do this now than later, while I still have fresh in my mind the code written for the main branch
 - Restored FSAA (this bug affected main branch as well, but another minor bug in 2.2 branch prevented the fixes from main branch from working). Also went for explicit resolve route which was planned from the start, so that VRAM saving feature could be implemented.
 - Save VRAM if FSAA is used + no postprocessing. This optimization is exclusive to Ogre 2.2+
     - This feature deprecates another overload of `UpdateRenderPassChain`
 - Fix bug where changing background colour would not take immediate effect unless the compositor was rebuilt. See changes to `Ogre2RenderTarget::UpdateBackgroundColor`
 - Removed Clear plasses in favour of LoadActions, which are more efficient in modern GPUs

**I apologize for this monolith of a PR. There were many fixes and improvements to implement which relied on merging main -> 2.2 and this merge causes too many conflicts to easily split this into multiple smaller PRs.**

**All ign-rendering tests are _passing_.**

## Summary

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

ahcorde commented on ogre2/include/ignition/rendering/ogre2/Ogre2RenderTarget.hh:
```diff
@@ -128,12 +155,34 @@ namespace ignition
       /// \param[in] _mask Visibility mask
       public: virtual void SetVisibilityMask(uint32_t _mask);
 
+      /// \brief Deprecated. Use other overloads.
+      public: static IGN_DEPRECATED(5) void UpdateRenderPassChain(
+          Ogre::CompositorWorkspace *_workspace,
+          const std::string &_workspaceDefName,
+          const std::string &_baseNode, const std::string &_finalNode,
+          const std::vector<RenderPassPtr> &_renderPasses,
+          bool _recreateNodes);
```
can we remove this ? This PR is targeting ign-rendering6

ahcorde commented on ogre2/include/ignition/rendering/ogre2/Ogre2RenderTarget.hh:
```diff
@@ -241,8 +296,20 @@ namespace ignition
+      /// Just in case we set this value to
+      /// Ogre2RenderTargetPrivate::ogreTexture[1] which is what most client
+      /// applications may want.
+      protected: IGN_DEPRECATED(5) Ogre::TextureGpu * ogreTexture = nullptr;
```
can we remove this ? This PR is targeting ign-rendering6

ahcorde commented on ogre2/src/Ogre2DepthCamera.cc:
```diff
@@ -297,11 +296,14 @@ void Ogre2DepthCamera::Destroy()
 
   // remove depth texture, material, compositor
-  if (this->dataPtr->ogreDepthTexture)
+  for( size_t i = 0u; i < 2u; ++i )
```
```suggestion
  for (size_t i = 0u; i < 2u; ++i)
```

darksylinc commented:
I've updated the PR with the suggested changes and removed code deprecated for ign-rendering5.

Now DCO bot says my tweets are unverified though Github disagrees; DCO bot is expecting me to use my github handle as username, for some reason

iche033 reviewed (commented):
Did some quick tests with a few worlds in ign-gazebo and seems to be working fine.

The `ogre2_demo` couldn't start though. I'm getting:

```
terminate called after throwing an instance of 'Ogre::InvalidParametersException'
  what():  OGRE EXCEPTION(2:InvalidParametersException): RenderTexture 'scene::RenderTexture(65435)0' must be resident! in GL3PlusRenderPassDescriptor::updateColourFbo at /home/osrf/code/ign-ogre_ws/src/ogre-next/RenderSystems/GL3Plus/src/OgreGL3PlusRenderPassDescriptor.cpp (line 223)
Aborted (core dumped)
```

darksylinc commented:
> Did some quick tests with a few worlds in ign-gazebo and seems to be working fine.
> 
> The `ogre2_demo` couldn't start though. I'm getting:
> 
> ```
> terminate called after throwing an instance of 'Ogre::InvalidParametersException'
>   what():  OGRE EXCEPTION(2:InvalidParametersException): RenderTexture 'scene::RenderTexture(65435)0' must be resident! in GL3PlusRenderPassDescriptor::updateColourFbo at /home/osrf/code/ign-ogre_ws/src/ogre-next/RenderSystems/GL3Plus/src/OgreGL3PlusRenderPassDescriptor.cpp (line 223)
> Aborted (core dumped)
> ```

Fixed. That slipped through my testing. I had to undo the VRAM saving code (reasons explained in the commit). It's just too much over-engineering to be worth it.

I thought it'd be easier as in, just flip a texture off and reroute a few nodes.

iche033 reviewed (approved):
works well for me now! I think windows CI is not yet working for ogre 2.2. Otherwise, other builds look good.
